File: lib/domain/agent/loop_detector.dart
import 'dart:convert';

import 'package:aios/domain/agent/tool_result.dart';

sealed class LoopCheckResult {
  const LoopCheckResult();
}

class LoopOk extends LoopCheckResult {
  const LoopOk();
}

class LoopWarning extends LoopCheckResult {
  const LoopWarning(this.count, this.toolName);
  final int count;
  final String toolName;
}

class LoopForceBreak extends LoopCheckResult {
  const LoopForceBreak();
}

class LoopDetector {
  static const _tag = 'AIOS-LoopDetector';

  final _loopOverrides = <String, Set<String>>{
    'screen_action': {'scroll', 'swipe', 'global'},
  };

  final List<({String tool, String argsCanonical})> _actionHistory = [];
  final List<String> _observationHistory = [];
  bool _warningGiven = false;

  void reset() {
    _actionHistory.clear();
    _observationHistory.clear();
    _warningGiven = false;
  }

  LoopCheckResult record(String toolName, String args, ToolResult result) {
    final canonical = _canonicalizeArgs(args);
    final observation = result.toContent();
    _actionHistory.add((tool: toolName, argsCanonical: canonical));
    _observationHistory.add(observation);

    final recentActions = _actionHistory.length >= 3
        ? _actionHistory.sublist(_actionHistory.length - 3)
        : _actionHistory.toList();

    final consecutiveDuplicates = recentActions
        .where((a) => a.tool == toolName && a.argsCanonical == canonical)
        .length;

    final isRepeatedAction =
        consecutiveDuplicates >= 3 &&
        !_isActionAllowedRepeated(toolName, canonical);

    final consecutiveIdenticalObs =
        _observationHistory.length >= 2 &&
        _observationHistory
                .sublist(_observationHistory.length - 2)
                .toSet()
                .length <
            2;

    if (isRepeatedAction || consecutiveIdenticalObs) {
      if (_warningGiven) {
        return const LoopForceBreak();
      }
      _warningGiven = true;
      return LoopWarning(consecutiveDuplicates, toolName);
    }

    return const LoopOk();
  }

  bool shouldNudge(int iteration, bool hasAnswer) {
    return iteration >= 3 && !hasAnswer && !_warningGiven;
  }

  String _canonicalizeArgs(String args) {
    try {
      final decoded = jsonDecode(args);
      if (decoded is! Map<String, dynamic>) return args;
      final keys = decoded.keys.toList()..sort();
      return keys.map((k) => '$k=${decoded[k]}').join(',');
    } on Object catch (e) {
      print('[$_tag] WARN: canonicalizeArgs error: $e');
      return args;
    }
  }

  bool _isActionAllowedRepeated(String tool, String argsCanonical) {
    final overrides = _loopOverrides[tool];
    if (overrides == null) return false;
    return overrides.any((o) => argsCanonical.contains(o));
  }
}

File: lib/src/agent/loop-detector.ts
import type { ToolResult } from '../types/agent';
import { toContent } from '../types/agent';

export type LoopCheckResult = LoopOk | LoopWarning | LoopForceBreak;

export interface LoopOk { type: 'ok' }
export interface LoopWarning { type: 'warning'; count: number; toolName: string }
export interface LoopForceBreak { type: 'forceBreak' }

export class LoopDetector {
  private actionHistory: Array<{ tool: string; argsCanonical: string }> = [];
  private observationHistory: string[] = [];
  private warningGiven = false;

  private loopOverrides: Record<string, string[]> = {
    screen_action: ['scroll', 'swipe', 'global'],
  };

  reset(): void {
    this.actionHistory = [];
    this.observationHistory = [];
    this.warningGiven = false;
  }

  record(toolName: string, args: string, result: ToolResult): LoopCheckResult {
    const canonical = this.canonicalizeArgs(args);
    const observation = toContent(result);
    this.actionHistory.push({ tool: toolName, argsCanonical: canonical });
    this.observationHistory.push(observation);

    const recentActions = this.actionHistory.slice(-3);
    const consecutiveDuplicates = recentActions.filter(
      (a) => a.tool === toolName && a.argsCanonical === canonical,
    ).length;

    const isRepeatedAction =
      consecutiveDuplicates >= 3 && !this.isActionAllowedRepeated(toolName, canonical);

    const consecutiveIdenticalObs =
      this.observationHistory.length >= 2 &&
      new Set(this.observationHistory.slice(-2)).size < 2;

    if (isRepeatedAction || consecutiveIdenticalObs) {
      if (this.warningGiven) {
        return { type: 'forceBreak' };
      }
      this.warningGiven = true;
      return { type: 'warning', count: consecutiveDuplicates, toolName };
    }

    return { type: 'ok' };
  }

  private canonicalizeArgs(args: string): string {
    try {
      const decoded = JSON.parse(args);
      if (typeof decoded !== 'object' || decoded === null) return args;
      const obj = decoded as Record<string, unknown>;
      return Object.keys(obj)
        .sort()
        .map((k) => `${k}=${obj[k]}`)
        .join(',');
    } catch {
      return args;
    }
  }

  private isActionAllowedRepeated(tool: string, argsCanonical: string): boolean {
    const overrides = this.loopOverrides[tool];
    if (!overrides) return false;
    return overrides.some((o) => argsCanonical.includes(o));
  }
}
